crates/red_knot/src/types.rs

```diff
@@ -37,40 +38,137 @@ impl TypeEnvironment {
     }
 }
 
+#[derive(Debug)]
 pub(crate) enum Type {
-    Class(ClassType),
+    // a specific function
     Function(FunctionType),
+    // the set of Python objects with this class in their MRO
+    Instance(ClassType),
+    Union(UnionType),
+    Intersection(IntersectionType),
+    // the dynamic or gradual type: a statically-unknown set of values
+    Any,
+    // the empty set of values
+    Never,
+    // unknown type (no annotation)
+    // equivalent to Any, or to object in strict mode
+    Unknown,
+    // name is not bound to any value
+    Unbound,
+    // TODO protocols, callable types, overloads, generics, type vars
 }
 
 impl Type {
-    pub(crate) fn name(&self) -> &str {
+    pub(crate) fn name(&self) -> Option<&str> {
         match self {
-            Type::Class(inner) => inner.name(),
-            Type::Function(inner) => inner.name(),
+            Type::Instance(inner) => Some(inner.name()),
+            Type::Function(inner) => Some(inner.name()),
+            Type::Union(_) => None,
+            Type::Intersection(_) => None,
+            Type::Any => Some("Any"),
+            Type::Never => Some("Never"),
+            Type::Unknown => Some("Unknown"),
+            Type::Unbound => Some("Unbound"),
+        }
+    }
+
+    fn display(&self, f: &mut std::fmt::Formatter, env: &TypeEnvironment) -> std::fmt::Result {
+        if let Some(name) = self.name() {
+            write!(f, "{}", name)
+        } else {
+            match self {
+                Type::Union(inner) => inner.display(f, env),
+                Type::Intersection(inner) => inner.display(f, env),
+                _ => Err(std::fmt::Error),
+            }
         }
     }
```

Comment:
It's a bit more annoying. But a common pattern here (see e.g. `Path`) is to have a `DisplayType` struct that implements `std::fmt::Display` and have the `display` function return that struct. It has the advantage that you can you the `display` function in `format_args` places (`write`, `format` `tracing` etc macros).

```suggestion
    fn display(&self, env: &TypeEnvironment) -> DisplayType {
        DisplayType { ty: self, environment: env }
    }
    
    #[derive(Copy, Clone, Debug)]
    struct DisplayType<'a> {
    	ty: &'a Type,
    	environment: &'a TypeEnvironment
  	}
  	
  	impl std::fmt::Display for DisplayType<'_> {
  		fn fmt(&self, &mut std::fmt::Formatter) -> std::fmt::Result {
  			if let Some(name) = self.ty.name() {
            write!(f, "{}", name)
        } else {
            match self {
                Type::Union(inner) => inner.display(f, env),
                Type::Intersection(inner) => inner.display(f, env),
                _ => Err(std::fmt::Error),
            }
        }			}
		}
```

Reply:
This makes a ton of sense, thank you!

You don't have to take the time to write out the code for a suggestion like this, unless you want to :) It's clear enough from the text.